File: code/c/dynamic-memory/array-db-for-chap.c
/* Program: small-db-2.c - array version */
#include <stdio.h>
#include <strings.h>
#include <stdbool.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>

// =====================
// = Type declarations =
// =====================

// The user can choose to add, delete, or print data or to quit
typedef enum {
    ADD_DATA,
    DELETE_DATA,
    PRINT_DATA,
    QUIT
} menu_option;

typedef union { /* same as original version */ } column_value;
typedef enum { /* same as original version */ } data_kind;
typedef struct { /* same as original version */ } row;

// The data store is a dynamic array of rows, keeping track
// of the number of rows in the array, and the id for the next row
typedef struct {
    int next_row_id;    // The id of the row that will be added next
    int row_count;      // The number of rows in the array
    row *rows;          // A pointer to the rows in memory
} data_store;

// ====================================
// = General Functions and Procedures =
// ====================================
void trim(char* text, int n) { /* same as original version */ }
bool is_integer(const char* text) { /* same as original version */ }
bool is_double(const char* text) { /* same as original version */ }
void clear_input() { /* same as original version */ }

// =====================================
// = Small DB Functions and Procedures =
// =====================================
row read_row(int next_id) { /* same as original version */ }
void print_row(row to_print) { /* same as original version */ }
menu_option get_menu_option() { /* same as original version */ }

// Add a row to the data store
void add_a_row(data_store *db_data)
{
    int row_id = 0;
    
    if (db_data == NULL) return;
    
    // Allocate the id
    row_id = db_data->next_row_id;
    db_data->next_row_id++;
    
    // Store the data - using realloc to allocate space for the rows
    db_data->row_count++;
    db_data->rows = (row *) realloc(db_data->rows, sizeof(row) * db_data->row_count);
    db_data->rows[db_data->row_count - 1] = read_row(row_id); // last idx = n - 1
}

// Get the array index of the row with the indicated ID
int idx_of_row_with_id(const data_store *db_data, int row_id)
{
    int i;
    
    if (db_data == NULL) return -1;
    
    // Loop through each element of the array
    for(i = 0; i < db_data->row_count; i++)
    {
        // is this the one we are after?
        if (db_data->rows[i].id == row_id)
        {
            return i; // return the index found.
        }
    }
    
    // Nothing found...
    return -1;
}

// Delete a row from the data store
void delete_a_row(data_store *db_data)
{
    int row_id, i, row_index;
    
    printf("Please enter id of row to delete: ");
    scanf("%d", &row_id);
    
    // Get the index of the row (will test if db_data == NULL)
    row_index = idx_of_row_with_id(db_data, row_id);
    
    if (row_index >= 0) // a row was found to delete...
    {
        // copy all data past the row, back over the row to delete it
        for(i = row_index; i < db_data->row_count - 1; i++)
        {
            // copy data back one spot in the array (one element at a time)
            db_data->rows[i] = db_data->rows[i+1];
        }
        
        // change the row count, and resize the array
        db_data->row_count--;
        db_data->rows = realloc(db_data->rows, sizeof(row) * db_data->row_count);
    }
}

// Print all of the rows from the data store
void print_all_rows(const data_store *db_data)
{
    int i = 0;
    
    if (db_data == NULL) return;
    
    // For each row in the array
    for (i = 0; i < db_data->row_count; i++)
    {
        // Print the row to the Terminal
        print_row(db_data->rows[i]);
    }
}


// ========
// = Main =
// ========

// Entry point
int main()
{
    menu_option opt;
    data_store db_data = {0, 0, NULL}; 
    
    do
    {
        opt = get_menu_option();
        
        switch(opt)
        {
            case ADD_DATA:
                add_a_row(&db_data);
                break;
            case DELETE_DATA:
                delete_a_row(&db_data);
                break;
            case PRINT_DATA:
                print_all_rows(&db_data);
                break;
            case QUIT:
                printf("Bye.\n");
                break;
        }
    } while(opt != QUIT);
}

File: code/pascal/dynamic-memory/ArrayDBforChap.pas
program SmallDb;
uses SysUtils;

// =====================
// = Type declarations =
// =====================

type 
    DataKind = {same as original version}
    ColumnValue = {same as original version}
    Row = {same as original version}
    
    // The data store is a dynamic linked list of rows, keeping track
    // of the number of rows in the list, and the id for the next row
    DataStore = record
        nextRowId: Integer;         // The id of the row that will be added next
        rows:      array of Row;    // A dynamic array of rows
    end;
    
    MenuOption = ( ADD_DATA, DELETE_DATA, PRINT_DATA, QUIT );


// =====================================
// = Small DB Functions and Procedures =
// =====================================

function ReadRow(nextId: Integer): Row; {same as original version}
procedure PrintRow(toPrint: row); {same as original version}

function GetMenuOption() : MenuOption;
var
    input: Integer = 0;
begin
    WriteLn('=========================');
    WriteLn('| Small DB              |');
    WriteLn('=========================');
    WriteLn(' 1: Add Data');
    WriteLn(' 2: Print Data');
    WriteLn(' 3: Delete Data');
    WriteLn(' 4: Quit');
    WriteLn('=========================');
    
    Write('Choose Option: ');
    ReadLn(input);
    
    while (input < 1) or (input > 4) do
    begin
        WriteLn('Please enter a value between 1 and 4.');
        Write('Choose Option: ');
        ReadLn(input)
    end;
    
    case input of
        1: result := ADD_DATA;
        2: result := PRINT_DATA;
        3: result := DELETE_DATA;
        4: result := QUIT;
        else result := QUIT;
    end;
end;

procedure AddRow(var dbData: DataStore);
var
    rowId: Integer = 0;
begin
    // Allocate the id
    rowId := dbData.nextRowId;
    dbData.nextRowId += 1;
    
    // Store the data
    SetLength(dbData.rows, Length(dbData.rows) + 1);
    dbData.rows[High(dbData.rows)] := ReadRow(rowId);
end;

function IndexOfRowWithID(const dbData: DataStore; rowId: Integer) : Integer;
var
    i: Integer;
begin
    // Loop through each element of the array
    for i := Low(dbData.rows) to High(dbData.rows) do
    begin
        // is this the one we are after?
        if dbData.rows[i].id = rowId then
        begin
            result := i;    // return the index found.
            exit;           // exit the function
        end;
    end;
    
    // Nothing found...
    result := -1;
end;

procedure DeleteRow(var dbData: DataStore);
var
    rowId, i, rowIndex: Integer;
begin
    WriteLn('Please enter id of row to delete: ');
    ReadLn(rowId);
    
    // Get the index of the row
    rowIndex := IndexOfRowWithID(dbData, rowId);
    
    if rowIndex >= 0 then // a row was found to delete...
    begin
        // copy all data past the row, back over the row to delete it
        for i := rowIndex to High(dbData.rows) - 1 do
        begin
            // copy data back one spot in the array (one element at a time)
            dbData.rows[i] := dbData.rows[i+1];
        end;
        
        // resize the array
        SetLength(dbData.rows, Length(dbData.rows) - 1);
    end;
end;

// Print all of the rows from the data store
procedure PrintAllRows(const dbData: DataStore);
var
    i: Integer = 0;
begin
    // For each row in the array
    for i := Low(dbData.rows) to High(dbData.rows) do
    begin
        // Print the row to the Terminal
        PrintRow(dbData.rows[i]);
    end;
end;

// ========
// = Main =
// ========

procedure Main();
var
    opt: MenuOption;
    dbData: DataStore; 
begin
    dbData.nextRowId := 0;
    SetLength(dbData.rows, 0);
    
    repeat
        opt := GetMenuOption();
        
        case opt of
            ADD_DATA:       AddRow(dbData);
            DELETE_DATA:    DeleteRow(dbData);
            PRINT_DATA:     PrintAllRows(dbData);
            QUIT:           WriteLn('Bye.');
        end;
    until opt = QUIT;
end;

begin
    Main();
end.